File: backend/tokens.py
"""Signierte Zugangstokens mit TTL — stateless (HMAC-SHA256, nur stdlib).

Ein Token kodiert `{sub, iat, exp}` und ist mit `AUTH_SECRET` signiert. Keine Datenbank,
keine Revocation-Liste (für 1–3 Nutzer ausreichend; Widerruf via Secret-Rotation).

CLI:
    python -m backend.tokens create --name philipp --days 90
    python -m backend.tokens verify <token>
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time


class TokenError(Exception):
    pass


def _b64e(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode()


def _b64d(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def _sign(secret: str, body: str) -> str:
    return _b64e(hmac.new(secret.encode(), body.encode(), hashlib.sha256).digest())


def mint(secret: str, name: str, ttl_seconds: int) -> str:
    now = int(time.time())
    payload = {"sub": name, "iat": now, "exp": now + int(ttl_seconds)}
    body = _b64e(json.dumps(payload, separators=(",", ":"), sort_keys=True).encode())
    return f"{body}.{_sign(secret, body)}"


def verify(secret: str, token: str) -> dict:
    if not secret:
        raise TokenError("AUTH_SECRET nicht gesetzt")
    try:
        body, sig = token.split(".", 1)
    except ValueError:
        raise TokenError("Token-Format ungültig")
    if not hmac.compare_digest(sig, _sign(secret, body)):
        raise TokenError("Signatur ungültig")
    try:
        payload = json.loads(_b64d(body))
    except Exception:
        raise TokenError("Payload ungültig")
    if int(payload.get("exp", 0)) < time.time():
        raise TokenError("Token abgelaufen")
    return payload


def _main() -> None:
    import argparse
    import sys
    from datetime import datetime, timezone

    from backend.config import get_settings

    ap = argparse.ArgumentParser(prog="python -m backend.tokens")
    sub = ap.add_subparsers(dest="cmd", required=True)
    c = sub.add_parser("create", help="Neues Token erzeugen")
    c.add_argument("--name", required=True, help="Name/Empfänger des Tokens")
    c.add_argument("--days", type=float, default=90.0, help="Gültigkeit in Tagen (TTL)")
    v = sub.add_parser("verify", help="Token prüfen")
    v.add_argument("token")
    args = ap.parse_args()

    secret = get_settings().auth_secret
    if not secret:
        sys.exit("Fehler: AUTH_SECRET ist nicht gesetzt (in .env / Umgebung).")

    if args.cmd == "create":
        token = mint(secret, args.name, int(args.days * 86400))
        exp = datetime.fromtimestamp(json.loads(_b64d(token.split(".")[0]))["exp"], tz=timezone.utc)
        print(token)
        print(f"# name={args.name}  gültig bis {exp:%Y-%m-%d %H:%M UTC}", file=sys.stderr)
    else:
        try:
            payload = verify(secret, args.token)
        except TokenError as e:
            sys.exit(f"ungültig: {e}")
        print(json.dumps(payload))


if __name__ == "__main__":
    _main()

File: internal/tokens/tokens.go
// Package tokens erzeugt und prüft signierte, zustandslose Zugangstokens (HMAC-SHA256).
// Format kompatibel zur früheren Python-Variante: base64url(payload).base64url(hmac),
// payload-JSON mit sortierten Keys und kompakten Separatoren {"exp":..,"iat":..,"sub":..}.
package tokens

import (
	"crypto/hmac"     // HMAC = signierter Hash mit geheimem Schlüssel
	"crypto/sha256"   // die genutzte Hash-Funktion
	"encoding/base64" // URL-sichere Kodierung der Bytes -> Text
	"encoding/json"   // zum sicheren Quoten des Namens
	"errors"
	"fmt"
	"strings"
	"time"
)

// Claims sind die im Token kodierten Felder.
//
// GO-EINSTEIGER: Die Texte in Backticks hinter den Feldern sind "struct tags".
// Sie sind Metadaten, die das json-Paket per Reflection ausliest: `json:"sub"`
// heißt "dieses Feld heißt im JSON 'sub'". Ohne Tag würde Go den Feldnamen "Sub"
// verwenden. So bleibt das JSON-Format kompatibel zur alten Python-Version.
type Claims struct {
	Sub string `json:"sub"` // subject = Name/Empfänger
	Iat int64  `json:"iat"` // issued at = Ausstellungszeit (Unix-Sekunden)
	Exp int64  `json:"exp"` // expiry = Ablaufzeit (Unix-Sekunden)
}

// b64 ist die rohe URL-Variante von base64 (ohne "="-Padding). Eine Paket-weite
// Variable auf Modulebene; `var name = wert` deklariert sie (Typ wird abgeleitet).
var b64 = base64.RawURLEncoding

// sign berechnet die HMAC-Signatur über body mit dem geheimen secret.
func sign(secret, body string) string {
	// hmac.New braucht eine Hash-FUNKTION (sha256.New, ohne Klammern = die
	// Funktion selbst) und den Schlüssel als Bytes.
	// []byte(secret) wandelt einen String in ein Byte-Slice — in Go sind Strings
	// unveränderlich, Bytes nicht; viele Crypto-/IO-APIs arbeiten mit []byte.
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(body))                 // die zu signierenden Daten einspeisen
	return b64.EncodeToString(mac.Sum(nil)) // Sum(nil) liefert die rohe Signatur als []byte
}

// Mint erzeugt ein Token für name mit Gültigkeit ttl.
// (ttl = "time to live" = Lebensdauer, vom Typ time.Duration.)
func Mint(secret, name string, ttl time.Duration) string {
	now := time.Now().Unix() // aktueller Zeitpunkt als Unix-Sekunden (int64)
	// Reihenfolge exp,iat,sub = sortierte Keys (kompatibel zu Python json sort_keys).
	// Wir bauen das JSON bewusst von Hand (statt json.Marshal), um exakt diese
	// Reihenfolge und kompakte Form zu garantieren. jsonString(name) sorgt fürs
	// korrekte Escapen/Quoten des Namens.
	body := b64.EncodeToString([]byte(fmt.Sprintf(
		`{"exp":%d,"iat":%d,"sub":%s}`, now+int64(ttl.Seconds()), now, jsonString(name),
	)))
	// Token = base64(payload) + "." + base64(signatur). String-Verkettung mit +.
	return body + "." + sign(secret, body)
}

// jsonString liefert s als JSON-String inkl. Anführungszeichen und Escaping.
func jsonString(s string) string {
	b, _ := json.Marshal(s) // kann bei einem String nicht fehlschlagen -> Fehler ignorieren
	return string(b)
}

// Verify prüft Signatur und Ablauf und liefert die Claims.
// Rückgabe (Claims, error): bei Fehlern wird ein leeres Claims{} + Fehler geliefert.
func Verify(secret, token string) (Claims, error) {
	if secret == "" {
		return Claims{}, errors.New("AUTH_SECRET nicht gesetzt")
	}
	// Token am "." in body und sig zerlegen. ok=false, wenn kein "." vorkommt.
	body, sig, ok := strings.Cut(token, ".")
	if !ok {
		return Claims{}, errors.New("Token-Format ungültig")
	}
	// WICHTIG: hmac.Equal vergleicht in konstanter Zeit (schützt vor Timing-
	// Angriffen). Niemals Signaturen mit == vergleichen! Wir signieren den body
	// erneut und prüfen, ob das Ergebnis mit der mitgelieferten Signatur übereinstimmt.
	if !hmac.Equal([]byte(sig), []byte(sign(secret, body))) {
		return Claims{}, errors.New("Signatur ungültig")
	}
	raw, err := b64.DecodeString(body) // base64 zurück in JSON-Bytes
	if err != nil {
		return Claims{}, errors.New("Payload ungültig")
	}
	var c Claims // Nullwert: alle Felder leer/0
	// Unmarshal füllt die Struct anhand der json-Tags aus den Bytes (Pythons
	// json.loads + Zuordnung auf ein Objekt in einem Schritt). `&c` = Adresse von
	// c, damit Unmarshal hineinschreiben kann.
	if err := json.Unmarshal(raw, &c); err != nil {
		return Claims{}, errors.New("Payload ungültig")
	}
	if c.Exp < time.Now().Unix() {
		return Claims{}, errors.New("Token abgelaufen")
	}
	return c, nil
}
